provider/apple.go

```diff
@@ -0,0 +1,529 @@
+		}
+	}
+
+	data := url.Values{}
+	data.Set("client_id", ah.conf.ClientID)
+	data.Set("client_secret", ah.conf.clientSecret) // JWT signed with Apple private key
+	data.Set("code", code)
+	data.Set("redirect_uri", redirectURI) // redirect URL can't refer to localhost and must have trusted certificate and https protocol
+	data.Set("grant_type", "authorization_code")
+
+	client := http.Client{Timeout: time.Second * 5}
+	req, err := http.NewRequestWithContext(ctx, "POST", ah.endpoint.TokenURL, strings.NewReader(data.Encode()))
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("content-type", appleRequestContentType)
+	req.Header.Add("accept", AcceptJSONHeader)
+	req.Header.Add("user-agent", defaultUserAgent) // apple requires a user agent
+
+	res, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	// Trying to decode (unmarshal json) data of response
+	err = json.NewDecoder(res.Body).Decode(result)
+	if err != nil {
+		return errors.Wrap(err, "unmarshalling data from apple service response failed")
+	}
+
+	// If above operation done successfully checking a response code and error descriptions, if one exist.
+	// Apple service will response either 200 (OK) or 400 (any error).
+	if res.StatusCode != http.StatusOK || result.Error != "" {
+		return fmt.Errorf("apple token service error: %s", result.Error)
+	}
+
+	defer func() {
+		if err = res.Body.Close(); err != nil {
+			ah.L.Logf("[ERROR] close request body failed when get access token: %v", err)
+		}
+	}()
+
+	return err
+}
+
+// createClientSecret use for create the JWT client secret required to make requests to the Apple validation server.
+// for more details go to link: https://developer.apple.com/documentation/sign_in_with_apple/generate_and_validate_tokens#3262048
+func (ah *AppleHandler) createClientSecret() (string, error) {
+	if ah.conf.privateKey == nil {
+		return "", errors.New("private key can't be empty")
+	}
+	// Create the Claims
+	now := time.Now()
+	exp := now.Add(time.Second * 1800).Unix() // default value
+
+	if ah.conf.ClientSecretExpire != 0 {
+		exp = now.Add(time.Second * time.Duration(ah.conf.ClientSecretExpire)).Unix()
+	}
+
+	claims := &jwt.StandardClaims{
+		Issuer:    ah.conf.TeamID,
+		IssuedAt:  now.Unix(),
+		ExpiresAt: exp,
+		Audience:  "https://appleid.apple.com",
+		Subject:   ah.conf.ClientID,
+	}
+
+	tkn := jwt.NewWithClaims(jwt.SigningMethodES256, claims)
+	tkn.Header["alg"] = "ES256"
+	tkn.Header["kid"] = ah.conf.KeyID
+
+	return tkn.SignedString(ah.conf.privateKey)
+}
+
+func (ah *AppleHandler) parseUserData(user *token.User, jUser string) {
+
+	type UserData struct {
+		Name struct {
+			FirstName string `json:"firstName"`
+			LastName  string `json:"lastName"`
+		} `json:"name"`
+		Email string `json:"email"`
+	}
+
+	var userData UserData
+
+	// Catch error for log only. No need break flow if user name doesn't exist
+	err := json.Unmarshal([]byte(jUser), &userData)
+	if err != nil {
```

Comment:
nit: this likely can be combined to the single statement, i.e. if err := blah ...